File: services/pipeline/python/consumer.py
"""
Kafka consumer wrapper for the Nier pipeline.

This module provides a high-level, type-safe interface for consuming messages
from Kafka topics with support for JSON deserialization and reliable processing.
"""

import logging
import signal
import threading
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Callable, Dict, List, Optional, Type, TypeVar, Union

from confluent_kafka import Consumer, KafkaError, KafkaException, TopicPartition

from .config import KafkaConfig
from .producer import NierProducer
from .schemas import Alert, DetectionEvent, FrameMetadata

logger = logging.getLogger(__name__)

T = TypeVar("T")


@dataclass
class MessageMetadata:
    """Metadata about a received message."""
    topic: str
    partition: int
    offset: int
    key: Optional[bytes] = None
    timestamp: Optional[int] = None
    headers: Dict[str, str] = field(default_factory=dict)


@dataclass
class IncomingMessage:
    """A received message with payload and metadata."""
    payload: bytes
    metadata: MessageMetadata

    def decode_json(self, cls: Type[T]) -> T:
        """
        Deserialize the payload as a schema object.

        Args:
            cls: Schema class to deserialize to (DetectionEvent, FrameMetadata, Alert)

        Returns:
            Deserialized object
        """
        return cls.from_bytes(self.payload)

    def decode_detection_event(self) -> DetectionEvent:
        """Deserialize as a DetectionEvent."""
        return DetectionEvent.from_bytes(self.payload)

    def decode_frame_metadata(self) -> FrameMetadata:
        """Deserialize as FrameMetadata."""
        return FrameMetadata.from_bytes(self.payload)

    def decode_alert(self) -> Alert:
        """Deserialize as an Alert."""
        return Alert.from_bytes(self.payload)

    @property
    def key_str(self) -> Optional[str]:
        """Get the message key as a string."""
        if self.metadata.key:
            return self.metadata.key.decode("utf-8")
        return None

    def header(self, key: str) -> Optional[str]:
        """Get a header value."""
        return self.metadata.headers.get(key)

    @property
    def correlation_id(self) -> Optional[str]:
        """Get the correlation ID header."""
        return self.header("correlation-id")

    @property
    def message_type(self) -> Optional[str]:
        """Get the message type header."""
        return self.header("message-type")


class MessageHandler(ABC):
    """Abstract base class for message handlers."""

    @abstractmethod
    def handle(self, message: IncomingMessage) -> None:
        """
        Process a single message.

        Args:
            message: Incoming message to process

        Raises:
            Exception: If processing fails
        """
        pass

    def on_error(self, message: IncomingMessage, error: Exception) -> None:
        """
        Called when message processing fails.

        Args:
            message: Message that failed to process
            error: Exception that occurred
        """
        logger.warning(
            "Message processing failed for topic=%s, partition=%d, offset=%d: %s",
            message.metadata.topic,
            message.metadata.partition,
            message.metadata.offset,
            error,
        )


class FunctionHandler(MessageHandler):
    """Wrapper to use a function as a message handler."""

    def __init__(self, func: Callable[[IncomingMessage], None]):
        self._func = func

    def handle(self, message: IncomingMessage) -> None:
        self._func(message)


class ConsumerError(Exception):
    """Exception raised for consumer errors."""
    pass


class NierConsumer:
    """High-level Kafka consumer wrapper for the Nier pipeline."""

    def __init__(
        self,
        config: KafkaConfig,
        dlq_producer: Optional[NierProducer] = None,
    ):
        """
        Create a new consumer with the given configuration.

        Args:
            config: Kafka configuration
            dlq_producer: Optional producer for dead letter queue
        """
        self.config = config
        self._consumer = Consumer(config.to_consumer_config())
        self._dlq_producer = dlq_producer
        self._running = False
        self._shutdown_event = threading.Event()

        logger.info(
            "Created Kafka consumer for %s with group %s",
            config.bootstrap_servers,
            config.consumer.group_id,
        )

    def subscribe(self, topics: List[str]) -> None:
        """
        Subscribe to the specified topics.

        Args:
            topics: List of topic names to subscribe to
        """
        logger.info("Subscribing to topics: %s", topics)
        self._consumer.subscribe(topics)

    def subscribe_all(self) -> None:
        """Subscribe to all Nier pipeline topics."""
        self.subscribe([
            self.config.topics.frames,
            self.config.topics.detections,
            self.config.topics.alerts,
        ])

    def subscribe_frames(self) -> None:
        """Subscribe to frame metadata topic."""
        self.subscribe([self.config.topics.frames])

    def subscribe_detections(self) -> None:
        """Subscribe to detection events topic."""
        self.subscribe([self.config.topics.detections])

    def subscribe_alerts(self) -> None:
        """Subscribe to alerts topic."""
        self.subscribe([self.config.topics.alerts])

    def commit(self) -> None:
        """Commit the current offsets synchronously."""
        self._consumer.commit()

    def commit_async(self) -> None:
        """Commit offsets asynchronously."""
        self._consumer.commit(asynchronous=True)

    def shutdown(self) -> None:
        """Signal shutdown to stop consuming."""
        logger.info("Signaling consumer shutdown")
        self._running = False
        self._shutdown_event.set()

    def run(
        self,
        handler: Union[MessageHandler, Callable[[IncomingMessage], None]],
        poll_timeout: float = 1.0,
    ) -> None:
        """
        Start consuming messages and process them with the given handler.

        Args:
            handler: Message handler or callback function
            poll_timeout: Timeout for each poll in seconds
        """
        if callable(handler) and not isinstance(handler, MessageHandler):
            handler = FunctionHandler(handler)

        self._running = True
        self._shutdown_event.clear()

        # Set up signal handlers for graceful shutdown
        original_sigint = signal.getsignal(signal.SIGINT)
        original_sigterm = signal.getsignal(signal.SIGTERM)

        def signal_handler(signum, frame):
            logger.info("Received signal %d, shutting down...", signum)
            self.shutdown()

        signal.signal(signal.SIGINT, signal_handler)
        signal.signal(signal.SIGTERM, signal_handler)

        logger.info("Starting message consumption loop")

        try:
            while self._running:
                msg = self._consumer.poll(poll_timeout)

                if msg is None:
                    continue

                if msg.error():
                    if msg.error().code() == KafkaError._PARTITION_EOF:
                        logger.debug(
                            "Reached end of partition %s [%d]",
                            msg.topic(),
                            msg.partition(),
                        )
                    else:
                        logger.error("Consumer error: %s", msg.error())
                    continue

                # Convert to IncomingMessage
                incoming = self._convert_message(msg)

                logger.debug(
                    "Received message from topic=%s, partition=%d, offset=%d",
                    incoming.metadata.topic,
                    incoming.metadata.partition,
                    incoming.metadata.offset,
                )

                try:
                    handler.handle(incoming)

                    if not self.config.consumer.enable_auto_commit:
                        self.commit_async()

                except Exception as e:
                    logger.error("Message processing failed: %s", e)
                    handler.on_error(incoming, e)

                    # Send to DLQ if configured
                    if self._dlq_producer:
                        try:
                            self._dlq_producer.send_to_dlq(
                                incoming.metadata.topic,
                                incoming.payload,
                                str(e),
                            )
                        except Exception as dlq_error:
                            logger.error("Failed to send to DLQ: %s", dlq_error)

        finally:
            # Restore original signal handlers
            signal.signal(signal.SIGINT, original_sigint)
            signal.signal(signal.SIGTERM, original_sigterm)

            # Final commit before shutdown
            if not self.config.consumer.enable_auto_commit:
                try:
                    self.commit()
                except Exception as e:
                    logger.warning("Failed to commit on shutdown: %s", e)

            logger.info("Consumer stopped")

    def run_with_callback(
        self,
        callback: Callable[[IncomingMessage], None],
        poll_timeout: float = 1.0,
    ) -> None:
        """
        Consume messages with a simple callback function.

        Args:
            callback: Function to call for each message
            poll_timeout: Timeout for each poll in seconds
        """
        self.run(FunctionHandler(callback), poll_timeout)

    def _convert_message(self, msg: Any) -> IncomingMessage:
        """Convert a confluent-kafka message to IncomingMessage."""
        headers = {}
        if msg.headers():
            for key, value in msg.headers():
                if value is not None:
                    headers[key] = value.decode("utf-8")

        return IncomingMessage(
            payload=msg.value() or b"",
            metadata=MessageMetadata(
                topic=msg.topic(),
                partition=msg.partition(),
                offset=msg.offset(),
                key=msg.key(),
                timestamp=msg.timestamp()[1] if msg.timestamp()[0] != 0 else None,
                headers=headers,
            ),
        )

    def assignment(self) -> List[TopicPartition]:
        """Get the current partition assignment."""
        return self._consumer.assignment()

    def position(self, partitions: List[TopicPartition]) -> List[TopicPartition]:
        """Get the current position for partitions."""
        return self._consumer.position(partitions)

    def pause(self, partitions: List[TopicPartition]) -> None:
        """Pause consumption for specific partitions."""
        self._consumer.pause(partitions)

    def resume(self, partitions: List[TopicPartition]) -> None:
        """Resume consumption for specific partitions."""
        self._consumer.resume(partitions)

    def close(self) -> None:
        """Close the consumer."""
        self._consumer.close()
        logger.info("Consumer closed")

    def __enter__(self) -> "NierConsumer":
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.close()


class ConsumerBuilder:
    """Builder for creating consumers with custom settings."""

    def __init__(self, bootstrap_servers: str):
        """
        Create a new consumer builder.

        Args:
            bootstrap_servers: Comma-separated list of broker addresses
        """
        self.config = KafkaConfig(bootstrap_servers=bootstrap_servers)
        self._dlq_producer: Optional[NierProducer] = None

    def group_id(self, group_id: str) -> "ConsumerBuilder":
        """Set the consumer group ID."""
        self.config.consumer.group_id = group_id
        return self

    def client_id(self, client_id: str) -> "ConsumerBuilder":
        """Set the client ID."""
        self.config.client_id = client_id
        return self

    def auto_offset_reset(self, reset: str) -> "ConsumerBuilder":
        """Set auto offset reset behavior."""
        self.config.consumer.auto_offset_reset = reset
        return self

    def enable_auto_commit(self, enable: bool) -> "ConsumerBuilder":
        """Enable or disable auto commit."""
        self.config.consumer.enable_auto_commit = enable
        return self

    def with_dlq_producer(self, producer: NierProducer) -> "ConsumerBuilder":
        """Set the dead letter queue producer."""
        self._dlq_producer = producer
        return self

    def build(self) -> NierConsumer:
        """Build the consumer."""
        return NierConsumer(self.config, self._dlq_producer)

File: services/pipeline/src/consumer.rs
//! Kafka consumer wrapper for the Nier pipeline.
//!
//! This module provides a high-level, type-safe interface for consuming messages
//! from Kafka topics with support for protobuf deserialization and reliable processing.

use crate::config::KafkaConfig;
use crate::producer::{NierProducer, ProducerError};
use prost::Message;
use rdkafka::consumer::{Consumer, StreamConsumer};
use rdkafka::message::{Headers, Message as KafkaMessage};
use rdkafka::TopicPartitionList;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;
use thiserror::Error;
use tokio::sync::broadcast;
use tracing::{debug, error, info, instrument, warn};

/// Errors that can occur during message consumption
#[derive(Error, Debug)]
pub enum ConsumerError {
    #[error("Failed to create consumer: {0}")]
    CreationError(String),

    #[error("Failed to subscribe to topics: {0}")]
    SubscriptionError(String),

    #[error("Failed to deserialize message: {0}")]
    DeserializationError(String),

    #[error("Failed to commit offset: {0}")]
    CommitError(String),

    #[error("Consumer poll error: {0}")]
    PollError(String),

    #[error("Message processing error: {0}")]
    ProcessingError(String),

    #[error("Consumer shutdown")]
    Shutdown,
}

/// Metadata about a received message
#[derive(Debug, Clone)]
pub struct MessageMetadata {
    /// Topic the message was received from
    pub topic: String,
    /// Partition the message was received from
    pub partition: i32,
    /// Offset of the message in the partition
    pub offset: i64,
    /// Message key (if present)
    pub key: Option<Vec<u8>>,
    /// Timestamp of the message
    pub timestamp: Option<i64>,
    /// Message headers
    pub headers: HashMap<String, String>,
}

/// A received message with payload and metadata
#[derive(Debug, Clone)]
pub struct IncomingMessage {
    /// Raw message payload
    pub payload: Vec<u8>,
    /// Message metadata
    pub metadata: MessageMetadata,
}

impl IncomingMessage {
    /// Deserialize the payload as a protobuf message
    pub fn decode_proto<M: Message + Default>(&self) -> Result<M, ConsumerError> {
        M::decode(self.payload.as_slice())
            .map_err(|e| ConsumerError::DeserializationError(e.to_string()))
    }

    /// Deserialize the payload as JSON
    pub fn decode_json<T: serde::de::DeserializeOwned>(&self) -> Result<T, ConsumerError> {
        serde_json::from_slice(&self.payload)
            .map_err(|e| ConsumerError::DeserializationError(e.to_string()))
    }

    /// Get the message key as a string
    pub fn key_str(&self) -> Option<String> {
        self.metadata
            .key
            .as_ref()
            .and_then(|k| String::from_utf8(k.clone()).ok())
    }

    /// Get a header value
    pub fn header(&self, key: &str) -> Option<&str> {
        self.metadata.headers.get(key).map(|s| s.as_str())
    }

    /// Get the correlation ID header
    pub fn correlation_id(&self) -> Option<&str> {
        self.header("correlation-id")
    }

    /// Get the message type header
    pub fn message_type(&self) -> Option<&str> {
        self.header("message-type")
    }
}

/// Handler trait for processing messages
#[async_trait::async_trait]
pub trait MessageHandler: Send + Sync {
    /// Process a single message
    async fn handle(&self, message: IncomingMessage) -> Result<(), ConsumerError>;

    /// Called when message processing fails
    async fn on_error(&self, message: IncomingMessage, error: ConsumerError) {
        warn!(
            "Message processing failed for topic={}, partition={}, offset={}: {}",
            message.metadata.topic,
            message.metadata.partition,
            message.metadata.offset,
            error
        );
    }
}

/// Function-based message handler
pub struct FnHandler<F>
where
    F: Fn(IncomingMessage) -> Result<(), ConsumerError> + Send + Sync,
{
    handler: F,
}

impl<F> FnHandler<F>
where
    F: Fn(IncomingMessage) -> Result<(), ConsumerError> + Send + Sync,
{
    pub fn new(handler: F) -> Self {
        Self { handler }
    }
}

#[async_trait::async_trait]
impl<F> MessageHandler for FnHandler<F>
where
    F: Fn(IncomingMessage) -> Result<(), ConsumerError> + Send + Sync,
{
    async fn handle(&self, message: IncomingMessage) -> Result<(), ConsumerError> {
        (self.handler)(message)
    }
}

/// High-level Kafka consumer wrapper
pub struct NierConsumer {
    consumer: StreamConsumer,
    config: Arc<KafkaConfig>,
    shutdown_tx: broadcast::Sender<()>,
    dlq_producer: Option<Arc<NierProducer>>,
}

impl NierConsumer {
    /// Create a new consumer with the given configuration
    pub fn new(config: KafkaConfig) -> Result<Self, ConsumerError> {
        info!(
            "Creating Kafka consumer for {} with group {}",
            config.bootstrap_servers, config.consumer.group_id
        );

        let consumer_config = config.build_consumer_config();
        let consumer: StreamConsumer = consumer_config
            .create()
            .map_err(|e| ConsumerError::CreationError(e.to_string()))?;

        let (shutdown_tx, _) = broadcast::channel(1);

        Ok(Self {
            consumer,
            config: Arc::new(config),
            shutdown_tx,
            dlq_producer: None,
        })
    }

    /// Set the dead letter queue producer
    pub fn with_dlq_producer(mut self, producer: Arc<NierProducer>) -> Self {
        self.dlq_producer = Some(producer);
        self
    }

    /// Get the configuration
    pub fn config(&self) -> &KafkaConfig {
        &self.config
    }

    /// Subscribe to the specified topics
    pub fn subscribe(&self, topics: &[&str]) -> Result<(), ConsumerError> {
        info!("Subscribing to topics: {:?}", topics);
        self.consumer
            .subscribe(topics)
            .map_err(|e| ConsumerError::SubscriptionError(e.to_string()))
    }

    /// Subscribe to all Nier pipeline topics
    pub fn subscribe_all(&self) -> Result<(), ConsumerError> {
        let topics = [
            self.config.topics.frames.as_str(),
            self.config.topics.detections.as_str(),
            self.config.topics.alerts.as_str(),
        ];
        self.subscribe(&topics)
    }

    /// Subscribe to frame metadata topic
    pub fn subscribe_frames(&self) -> Result<(), ConsumerError> {
        self.subscribe(&[self.config.topics.frames.as_str()])
    }

    /// Subscribe to detection events topic
    pub fn subscribe_detections(&self) -> Result<(), ConsumerError> {
        self.subscribe(&[self.config.topics.detections.as_str()])
    }

    /// Subscribe to alerts topic
    pub fn subscribe_alerts(&self) -> Result<(), ConsumerError> {
        self.subscribe(&[self.config.topics.alerts.as_str()])
    }

    /// Commit the current offsets synchronously
    pub fn commit(&self) -> Result<(), ConsumerError> {
        self.consumer
            .commit_consumer_state(rdkafka::consumer::CommitMode::Sync)
            .map_err(|e| ConsumerError::CommitError(e.to_string()))
    }

    /// Commit offsets asynchronously
    pub fn commit_async(&self) {
        if let Err(e) = self
            .consumer
            .commit_consumer_state(rdkafka::consumer::CommitMode::Async)
        {
            warn!("Failed to commit offsets asynchronously: {}", e);
        }
    }

    /// Get a shutdown receiver
    pub fn shutdown_receiver(&self) -> broadcast::Receiver<()> {
        self.shutdown_tx.subscribe()
    }

    /// Signal shutdown to stop consuming
    pub fn shutdown(&self) {
        info!("Signaling consumer shutdown");
        let _ = self.shutdown_tx.send(());
    }

    /// Start consuming messages and process them with the given handler
    #[instrument(skip(self, handler))]
    pub async fn run<H: MessageHandler>(&self, handler: Arc<H>) -> Result<(), ConsumerError> {
        use rdkafka::message::Message as _;
        use tokio_stream::StreamExt;

        let mut shutdown_rx = self.shutdown_receiver();
        let stream = self.consumer.stream();
        tokio::pin!(stream);

        info!("Starting message consumption loop");

        loop {
            tokio::select! {
                _ = shutdown_rx.recv() => {
                    info!("Received shutdown signal");
                    break;
                }
                message_result = stream.next() => {
                    match message_result {
                        Some(Ok(borrowed_message)) => {
                            let incoming = self.convert_message(&borrowed_message);

                            debug!(
                                "Received message from topic={}, partition={}, offset={}",
                                incoming.metadata.topic,
                                incoming.metadata.partition,
                                incoming.metadata.offset
                            );

                            match handler.handle(incoming.clone()).await {
                                Ok(()) => {
                                    if !self.config.consumer.enable_auto_commit {
                                        self.commit_async();
                                    }
                                }
                                Err(e) => {
                                    error!("Message processing failed: {}", e);
                                    handler.on_error(incoming.clone(), e).await;

                                    // Send to DLQ if configured
                                    if let Some(ref dlq) = self.dlq_producer {
                                        if let Err(dlq_err) = dlq
                                            .send_to_dlq(
                                                &incoming.metadata.topic,
                                                &incoming.payload,
                                                "Processing failed",
                                            )
                                            .await
                                        {
                                            error!("Failed to send to DLQ: {}", dlq_err);
                                        }
                                    }
                                }
                            }
                        }
                        Some(Err(e)) => {
                            error!("Kafka error: {}", e);
                        }
                        None => {
                            debug!("Stream ended");
                            break;
                        }
                    }
                }
            }
        }

        // Final commit before shutdown
        if !self.config.consumer.enable_auto_commit {
            if let Err(e) = self.commit() {
                warn!("Failed to commit on shutdown: {}", e);
            }
        }

        Ok(())
    }

    /// Consume messages with a simple callback function
    pub async fn run_with_callback<F, Fut>(&self, callback: F) -> Result<(), ConsumerError>
    where
        F: Fn(IncomingMessage) -> Fut + Send + Sync + 'static,
        Fut: std::future::Future<Output = Result<(), ConsumerError>> + Send,
    {
        use rdkafka::message::Message as _;
        use tokio_stream::StreamExt;

        let mut shutdown_rx = self.shutdown_receiver();
        let stream = self.consumer.stream();
        tokio::pin!(stream);

        loop {
            tokio::select! {
                _ = shutdown_rx.recv() => {
                    break;
                }
                message_result = stream.next() => {
                    match message_result {
                        Some(Ok(borrowed_message)) => {
                            let incoming = self.convert_message(&borrowed_message);
                            if let Err(e) = callback(incoming).await {
                                error!("Callback error: {}", e);
                            } else if !self.config.consumer.enable_auto_commit {
                                self.commit_async();
                            }
                        }
                        Some(Err(e)) => {
                            error!("Kafka error: {}", e);
                        }
                        None => break,
                    }
                }
            }
        }

        Ok(())
    }

    /// Convert a borrowed Kafka message to our IncomingMessage type
    fn convert_message<M: KafkaMessage>(&self, msg: &M) -> IncomingMessage {
        let payload = msg.payload().unwrap_or(&[]).to_vec();
        let key = msg.key().map(|k| k.to_vec());

        let mut headers = HashMap::new();
        if let Some(h) = msg.headers() {
            for header in h.iter() {
                if let Some(value) = header.value {
                    if let Ok(v) = String::from_utf8(value.to_vec()) {
                        headers.insert(header.key.to_string(), v);
                    }
                }
            }
        }

        IncomingMessage {
            payload,
            metadata: MessageMetadata {
                topic: msg.topic().to_string(),
                partition: msg.partition(),
                offset: msg.offset(),
                key,
                timestamp: msg.timestamp().to_millis(),
                headers,
            },
        }
    }

    /// Get the current partition assignment
    pub fn assignment(&self) -> Result<TopicPartitionList, ConsumerError> {
        self.consumer
            .assignment()
            .map_err(|e| ConsumerError::PollError(e.to_string()))
    }

    /// Get the current position for all assigned partitions
    pub fn position(&self) -> Result<TopicPartitionList, ConsumerError> {
        self.consumer
            .position()
            .map_err(|e| ConsumerError::PollError(e.to_string()))
    }

    /// Pause consumption for specific partitions
    pub fn pause(&self, partitions: &TopicPartitionList) -> Result<(), ConsumerError> {
        self.consumer
            .pause(partitions)
            .map_err(|e| ConsumerError::PollError(e.to_string()))
    }

    /// Resume consumption for specific partitions
    pub fn resume(&self, partitions: &TopicPartitionList) -> Result<(), ConsumerError> {
        self.consumer
            .resume(partitions)
            .map_err(|e| ConsumerError::PollError(e.to_string()))
    }
}

/// Builder for creating consumers with custom settings
pub struct ConsumerBuilder {
    config: KafkaConfig,
    dlq_producer: Option<Arc<NierProducer>>,
}

impl ConsumerBuilder {
    /// Create a new consumer builder
    pub fn new(bootstrap_servers: impl Into<String>) -> Self {
        Self {
            config: KafkaConfig::new(bootstrap_servers),
            dlq_producer: None,
        }
    }

    /// Set the consumer group ID
    pub fn group_id(mut self, group_id: impl Into<String>) -> Self {
        self.config.consumer.group_id = group_id.into();
        self
    }

    /// Set the client ID
    pub fn client_id(mut self, client_id: impl Into<String>) -> Self {
        self.config.client_id = client_id.into();
        self
    }

    /// Set auto offset reset behavior
    pub fn auto_offset_reset(mut self, reset: impl Into<String>) -> Self {
        self.config.consumer.auto_offset_reset = reset.into();
        self
    }

    /// Enable or disable auto commit
    pub fn enable_auto_commit(mut self, enable: bool) -> Self {
        self.config.consumer.enable_auto_commit = enable;
        self
    }

    /// Set the dead letter queue producer
    pub fn with_dlq_producer(mut self, producer: Arc<NierProducer>) -> Self {
        self.dlq_producer = Some(producer);
        self
    }

    /// Build the consumer
    pub fn build(self) -> Result<NierConsumer, ConsumerError> {
        let mut consumer = NierConsumer::new(self.config)?;
        if let Some(dlq) = self.dlq_producer {
            consumer = consumer.with_dlq_producer(dlq);
        }
        Ok(consumer)
    }
}

/// Async trait for message handlers (re-export for convenience)
pub use async_trait::async_trait;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_incoming_message_headers() {
        let mut headers = HashMap::new();
        headers.insert("correlation-id".to_string(), "test-123".to_string());
        headers.insert("message-type".to_string(), "detection_event".to_string());

        let message = IncomingMessage {
            payload: vec![1, 2, 3],
            metadata: MessageMetadata {
                topic: "test".to_string(),
                partition: 0,
                offset: 100,
                key: Some(b"key".to_vec()),
                timestamp: Some(1234567890),
                headers,
            },
        };

        assert_eq!(message.correlation_id(), Some("test-123"));
        assert_eq!(message.message_type(), Some("detection_event"));
        assert_eq!(message.key_str(), Some("key".to_string()));
    }
}
